File: solutions/codeforces/676D.py
from Queue import Queue
import sys

def enc(a, r, c):
    return (a << 20) | (r << 10) | c

def dec(v):
    return v >> 20, (v >> 10) & 1023, v & 1023

ROTATE = {
    ord('+') : ord('+'),
    ord('-') : ord('|'), ord('|') : ord('-'),
    ord('^') : ord('>'), ord('>') : ord('v'), ord('v') : ord('<'), ord('<') : ord('^'),
    ord('L') : ord('U'), ord('U') : ord('R'), ord('R') : ord('D'), ord('D') : ord('L'),
    ord('*') : ord('*')
}

N, M = map(int, raw_input().split())
A = [[], [], [], []]
V = set()
for _ in xrange(N):
    A[0].append(map(ord, list(raw_input())))
    for i in xrange(1, 4):
        row = []
        for j in xrange(M):
            row.append(ROTATE[A[i-1][-1][j]])
        A[i].append(row)
SR, SC = map(lambda x: int(x)-1, raw_input().split())
ER, EC = map(lambda x: int(x)-1, raw_input().split())

DRC = [(0, -1), (0, 1), (-1, 0), (1, 0)]
LF = map(ord, list("+-<URD"))
RT = map(ord, list("+->LUD"))
UP = map(ord, list("+|^LRD"))
DO = map(ord, list("+|vLUR"))

ans = 0
Q = Queue()
Q.put(enc(0, SR, SC))
while not Q.empty():
    sz = Q.qsize()
    for _ in xrange(sz):
        a, r, c = rec = dec(Q.get())
        if rec in V:
            continue
        V.add(rec)
        if r == ER and c == EC:
            print ans
            sys.exit(0)
        Q.put(enc((a+1)%4, r, c))
        for dr, dc in DRC:
            nr = r+dr
            nc = c+dc
            if nr < 0 or nr >= N or nc < 0 or nc >= M or (a, nr, nc) in V:
                continue
            v1 = A[a][r][c]
            v2 = A[a][nr][nc]
            if (dc == -1 and v1 in LF and v2 in RT) or \
               (dc ==  1 and v1 in RT and v2 in LF) or \
               (dr == -1 and v1 in UP and v2 in DO) or \
               (dr ==  1 and v1 in DO and v2 in UP):
                Q.put(enc(a, nr, nc))
    ans += 1
print -1

File: solutions/codeforces/676D.cpp
#include <stdio.h>
#include <string.h>
#include <queue>
using namespace std;

int N, M;
int SR, SC, ER, EC;
char A[4][1010][1010];
bool V[4][1010][1010];
char R[256];
char left[] = "+-<URD";
char right[] = "+->LUD";
char up[] = "+|^LRD";
char down[] = "+|vLUR";

int enc(int a, int r, int c) { return (a << 20) | (r << 10) | c; }
int dec_a(int v) { return v >> 20; }
int dec_r(int v) { return (v >> 10) & 1023; }
int dec_c(int v) { return v & 1023; }

const int dr[] = {0, 0, -1, 1};
const int dc[] = {-1, 1, 0, 0};
int main() {
  R['+'] = '+';
  R['-'] = '|'; R['|'] = '-';
  R['^'] = '>'; R['>'] = 'v'; R['v'] = '<'; R['<'] = '^';
  R['L'] = 'U'; R['U'] = 'R'; R['R'] = 'D'; R['D'] = 'L';
  R['*'] = '*';
  scanf("%d%d", &N, &M);
  for (int i = 0; i < N; i++)
    scanf("%s", A[0][i]);
  scanf("%d%d", &SR, &SC);
  scanf("%d%d", &ER, &EC);
  SR--; SC--; ER--; EC--;
  for (int i = 1; i < 4; i++)
    for (int j = 0; j < N; j++)
      for (int k = 0; k < M; k++)
        A[i][j][k] = R[A[i-1][j][k]];

  queue<int> Q;
  int ans = 0;
  Q.push(enc(0, SR, SC));
  while (!Q.empty()) {
    int sz = Q.size();
    while (sz--) {
      int v = Q.front();
      Q.pop();
      int a = dec_a(v);
      int r = dec_r(v);
      int c = dec_c(v);
      if (V[a][r][c]) continue;
      V[a][r][c] = true;
      if (r == ER && c == EC) {
        printf("%d\n", ans);
        return 0;
      }
      Q.push(enc((a+1)%4, r, c));
      for (int i = 0; i < 4; i++) {
        int nr = r+dr[i];
        int nc = c+dc[i];
        if (nr < 0 || nr >= N || nc < 0 || nc >= M || V[a][nr][nc])
          continue;
        char v1 = A[a][r][c];
        char v2 = A[a][nr][nc];
        if ((i == 0 && strchr(left, v1) && strchr(right, v2)) ||
            (i == 1 && strchr(right, v1) && strchr(left, v2)) ||
            (i == 2 && strchr(up, v1) && strchr(down, v2)) ||
            (i == 3 && strchr(down, v1) && strchr(up, v2)))
          Q.push(enc(a, nr, nc));
      }
    }
    ans++;
  }

  printf("%d\n", -1);
  return 0;
}